File: ios/Sources/GutenbergKit/Sources/Model/HTTP/EditorURLResponse.swift
import Foundation

/// An HTTP response containing body data and headers.
///
/// This struct encapsulates the response from an HTTP request, storing both the
/// response body and headers. It's used throughout the editor for caching API
/// responses and building preload data.
///
/// The `description` property formats the response as JSON suitable for injection
/// into the editor's preload system.
public struct EditorURLResponse: Sendable, Equatable, Codable, Hashable {

    /// The response body data.
    let data: Data

    /// The HTTP response headers.
    let responseHeaders: EditorHTTPHeaders

    /// Creates a response from raw data and an `HTTPURLResponse`.
    ///
    /// - Parameters:
    ///   - data: The response body data.
    ///   - httpUrlResponse: The URL response containing headers.
    init(data: Data, httpUrlResponse: HTTPURLResponse) {
        self.data = data
        self.responseHeaders = EditorHTTPHeaders(httpUrlResponse.allHeaderFields)
    }

    /// Creates a response from raw data and headers.
    ///
    /// - Parameters:
    ///   - data: The response body data.
    ///   - responseHeaders: The HTTP headers.
    init(data: Data, responseHeaders: EditorHTTPHeaders) {
        self.data = data
        self.responseHeaders = responseHeaders
    }

    /// Creates a response from a string body and headers.
    ///
    /// The string is encoded as UTF-8 data.
    ///
    /// - Parameters:
    ///   - string: The response body as a string.
    ///   - responseHeaders: The HTTP headers.
    init(string: String, responseHeaders: EditorHTTPHeaders) {
        self.data = Data(string.utf8)
        self.responseHeaders = responseHeaders
    }

    /// Creates a response from a tuple of data and `HTTPURLResponse`.
    ///
    /// Convenience initializer for working with `URLSession` results.
    ///
    /// - Parameter rawResponse: A tuple containing the response data and URL response.
    init(_ rawResponse: (Data, HTTPURLResponse)) {
        self = EditorURLResponse(data: rawResponse.0, httpUrlResponse: rawResponse.1)
    }

    func toJSON() throws -> JSON {
        .object([
            "body": try JSON(self.data),
            "headers": try responseHeaders.toJSON()
        ])
    }

    /// An empty response with an empty JSON object body and no headers.
    static let empty: EditorURLResponse = EditorURLResponse(string: "{}", responseHeaders: [:])
}

File: android/Gutenberg/src/main/java/org/wordpress/gutenberg/model/http/EditorURLResponse.kt
package org.wordpress.gutenberg.model.http

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject

/**
 * An HTTP response containing body data and headers.
 *
 * This class encapsulates the response from an HTTP request, storing both the
 * response body and headers. It's used throughout the editor for caching API
 * responses and building preload data.
 */
@Serializable
data class EditorURLResponse(
    /** The response body data as a string. */
    val data: String,
    /** The HTTP response headers. */
    val responseHeaders: EditorHTTPHeaders
) {
    fun toJsonElement(): JsonElement {
        val bodyElement = try {
            Json.parseToJsonElement(data)
        } catch (e: Exception) {
            kotlinx.serialization.json.JsonPrimitive(data)
        }

        return JsonObject(
            mapOf(
                "body" to bodyElement,
                "headers" to responseHeaders.toJsonElement()
            )
        )
    }

    companion object {
        /** An empty response with an empty JSON object body and no headers. */
        val empty = EditorURLResponse(data = "{}", responseHeaders = EditorHTTPHeaders.empty)
    }
}

/**
 * Creates a copy of the response with only preload-relevant headers.
 *
 * Filters headers to only include those expected by WordPress core's preload system.
 */
fun EditorURLResponse.asPreloadResponse(): EditorURLResponse {
    // These headers were chosen because they're the same as those present in WP Core
    val headers = this.responseHeaders.filtering("Accept", "Link")
    return EditorURLResponse(data = this.data, responseHeaders = headers)
}
